[DCA] Update Gitlab pipeline for DCA releases.

devonboyer opened the pull request:
### What does this PR do?

Tags for DCA releases must start with `dca`, like `dca-1.0.0`.

- Add `releasenotes/cluster-agent` directory in root of project.

### Limitations/Hacks

- The proposed solution uses git tags that have the prefix `dca-` which means that the git tag doesn't conform to semantic versioning unless this prefix is ignored or removed.
- All jobs in the pipeline for the agent need to have an `except` condition to not run for DCA release tags since the omnibus build for the agent fails because the git tag doesn't conform to semantic versioning.
- Gitlab does not support running a job for a specific tag so the best we can do is exclude jobs for any git ref (branches or tags) matching the regex for dca releases.
- When creating `releasenotes` for the DCA you must use the `-d releasenotes/cluster-agent` flag for reno.
- The DCA image tag names will include the `dca-` prefix too.
### Motivation

Prepare for DCA 1.0.0 release.

devonboyer commented on tasks/cluster_agent.py:
```diff
@@ -123,3 +127,36 @@ def image_build(ctx):
+    ctx.run("git commit -m \"Add prelude for {} release\"".format(new_version))
+
+@task
+def update_changelog(ctx, new_version):
```
I'll take another look at `release.update_changelog`.

devonboyer replied:
The problem I had in duplicating the logic was that there is no previous release of the DCA and I wasn't sure what the challenges were with always re-generating the entire CHANGELOG every release?

hush-hush replied:
The issue is that we branch out for bugfix release. This means the, for example, the tags `6.3.1`, `6.3.2` and `6.3.3` don't exist on `master` but only on `6.3.x`. If you generate the entire changelog from master you will miss every bugfix milestone.

That's why we generate the diff between the last bugfix release and the new release (for example: diff between `6.3.3` an `6.4.0`).

devonboyer commented on releasenotes-dca/notes/add-changelog-3b1d5a12dd57a2b3.yaml:
```diff
@@ -0,0 +1,4 @@
+---
+enhancements:
```
This was just placeholder so the folder would get committed. I'll delete it.

devonboyer commented on CHANGELOG-DCA.rst:
```diff
@@ -0,0 +1,17 @@
+
+.. _Release Notes_dca-1.0.0:
+
+dca-1.0.0
```
This was just placeholder so CHANGELOG wasn't empty. I'll delete it.

devonboyer commented on .gitlab-ci.yml:
```diff
@@ -97,6 +97,22 @@ before_script:
+
+.run_on_cluster_agent_ref: &run_on_cluster_agent_ref
+  only:
+    - /^dca([\d.-]|rc)+$/
```
This regex can match any git ref, which includes branches or tags. Gitlab does not support running a job for a specific tag.

hush-hush replied:
I this you can. Look at those 2 doc https://docs.gitlab.com/ee/ci/yaml/#only-and-except-simplified and https://docs.gitlab.com/ee/ci/variables/#predefined-variables-environment-variables.

I'm guessing this will do the trick:
```yaml
only:
  refs:
    - tags
  variables:
    - CI_COMMIT_TAG =~ /^dca([\d.-]|rc)+$/
```

I didn't tested it ...

devonboyer replied:
That could work. It does say you can use regexes with variables. 👍

hush-hush commented on .gitlab-ci.yml:
```diff
@@ -968,24 +968,29 @@ latest_release:
-  <<: *run_when_triggered_on_tag
+  only:
+    refs:
+      - tags
```
I this you can. Look at those 2 doc https://docs.gitlab.com/ee/ci/yaml/#only-and-except-simplified and https://docs.gitlab.com/ee/ci/variables/#predefined-variables-environment-variables.

I'm guessing this will do the trick:
```yaml
only:
  refs:
    - tags
  variables:
    - CI_COMMIT_TAG =~ /^dca([\d.-]|rc)+$/
```

I didn't tested it ...

hush-hush commented on .gitlab-ci.yml:
```diff
@@ -538,6 +573,7 @@ deploy_windows_testing:
 
 # run dd-agent-testing on windows
 kitchen_windows:
+  <<: *skip_on_cluster_agent_tag
```
We skip stage `testkitchen_testing` and `testkitchen_cleanup` but not `testkitchen_deploy`. I think we should skip it too, especially since we're skipping the clean up.

devonboyer replied:
Since these jobs already have

```yaml
only:
    - triggers
```

they don't run when any tag is pushed so I don't think I need to add the `skip_on_cluster_agent_tag` here?

hush-hush commented on tasks/cluster_agent.py:
```diff
@@ -123,3 +127,38 @@ def image_build(ctx):
+
+
+@task
+def add_prelude(ctx, new_version):
```
Both `add_prelude` and `update_changelog` only work on linux. I don't know if it's a limitation we want to change, but we should add a comment in the docstring.

hush-hush commented on tasks/utils.py:
```diff
@@ -153,7 +153,7 @@ def query_version(ctx, git_sha_length=7):
     described_version = ctx.run("git describe --tags", hide=True).stdout.strip()
     # For the tag 6.0.0-beta.0, this will match 6.0.0
-    version_match = re.findall(r"^v?(\d+\.\d+\.\d+)", described_version)
+    version_match = re.findall(r"^(?:dca)?-?v?(\d+\.\d+\.\d+)", described_version)
```
I think it should be required, but no strong opinion here (just `dca-1.0.0` is easier to read than `dca1.0.0`).

hush-hush commented on tasks/cluster_agent.py:
```diff
@@ -123,3 +127,38 @@ def image_build(ctx):
+
+
+@task
+def update_changelog(ctx, new_version):
```
This doesn't handle bugfix release (like said in my other comment).

We're in the process of updating the `update_changelog` method for the agent to work on windows. Their is a lot of code duplication between `task/cluster_agent.py` and `task/release.py`. If you want we could merge both command set in another PR ?

devonboyer commented:
@hush-hush Can you take another look at this PR?

- I changed the regex to require the `-` in the `dca-` prefix
- I removed the invoke tasks for release notes. I will do this in another PR (ideally we just use the Agent's tasks for updating the CHANGELOG like you suggested)
- I didn't add `skip_on_cluster_agent_tag` for jobs that have `run_when_triggered` since they shouldn't run anyway.

Commit:
Add cluster agent conditions for gitlab jobs.

CharlyF reviewed (approved):
This looks good to me, although I'd love for someone from @DataDog/agent-core to validate this. 
We would merge after 6.4.2 is pushed.

hush-hush reviewed (approved):
Thanks for all the work on this @devonboyer !
